js/ui/tag_box.js

```diff
@@ -1195,12 +1195,20 @@ var TagBox = SelectBox.inherit({
     },
 
     _applyButtonHandler: function() {
-        this.option("value", this._getListValues());
+        this.option("value", this._getSortedListValues());
         this._clearTextValue();
         this._clearFilter();
         this.callBase();
     },
 
+    _getSortedListValues: function() {
+        var listValues = this._getListValues(),
+            currentValue = this.option("value"),
+            sortFunction = function(item1, item2) { return currentValue.indexOf(item2) - currentValue.indexOf(item1); };
```

Comment:
Values may be objects when valueExpr is "this". IndexOf may be incorrect. There is `this._isValueEquals` method in the dxTagBox which may be used to compare 2 value instances

Reply:
List's selectedItems contains the same objects that TagBox's value. I've added an additional test to cover scenario that you have described.